lib/accesslists/hierarchy.go

```diff
@@ -0,0 +1,569 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package accesslists
+
+import (
+	"context"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/types/accesslist"
+	"github.com/gravitational/teleport/lib/services"
+)
+
+// RelationshipKind represents the type of relationship: member or owner.
+type RelationshipKind int
+
+const (
+	RelationshipKindMember RelationshipKind = iota
+	RelationshipKindOwner
+)
+
+// MembershipOrOwnershipType represents the type of membership or ownership a User has for an Access List.
+type MembershipOrOwnershipType int
+
+const (
+	MembershipOrOwnershipTypeNone MembershipOrOwnershipType = iota
+	MembershipOrOwnershipTypeExplicit
+	MembershipOrOwnershipTypeInherited
+)
+
+type MembersAndLocksGetter interface {
+	ListAccessListMembers(ctx context.Context, accessListName string, pageSize int, pageToken string) (members []*accesslist.AccessListMember, nextToken string, err error)
+}
+
+// HierarchyNode represents an access list and its relationships.
+type HierarchyNode struct {
+	// AccessList is the Underlying AccessList object.
+	AccessList *accesslist.AccessList
+	// MemberUsers are users who are direct members of this list.
+	MemberUsers map[string]struct{}
+	// MemberLists are AccessLists that are direct members of this list.
+	MemberLists map[string]*HierarchyNode
+	// OwnerUsers are users who are direct Owners of this list.
+	OwnerUsers map[string]struct{}
+	// OwnerLists are AccessLists that are direct Owners of this list.
+	OwnerLists map[string]*HierarchyNode
+	// MemberParents are AccessLists that have this list as a member.
+	MemberParents map[string]*HierarchyNode
+	// OwnerParents are AccessLists that have this list as an owner.
+	OwnerParents map[string]*HierarchyNode
+}
+
+type Hierarchy struct {
```

Comment:
I would define `Hierarchy` interface so it's clear what it implements and rename this struct to `hierarchy`.